### PR TITLE
 SIL: Force Clang-imported protocol conformances to get deserialized when used.

### PR DESCRIPTION
**Pull request by jckarter**
Code may end up indirectly using a witness table for a Clang-imported type by inlining code that used the conformance from another module, in which case we need to ensure we have a local definition at hand in the inlining module so we can have something to link against independently. This needs to be fixed from both sides:

- During serialization, serialize not only witness tables from the current module, but from Clang-imported modules too
- During deserialization, when the SILLinker walks a loaded module, ensure that all shared conformances get deserialized, including those from ApplyInsts and inherited/associated type protocol requirements.

Fixes rdar://problem/38687726.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@swift-ci Please test compiler performance

**Comment by jckarter** [comment]
@slavapestov I went back to the eager deserialization approach since trying to do it lazily in IRGen is too late due to pre-IRGen lowering passes on SIL. How does this look (assuming the performance impact is acceptable)?

File: test/IRGen/Inputs/deserialize-clang-importer-witness-tables/clang-importer-witness-tables/regex.swift [hunk 1]
```diff
@@ -0,0 +1 @@
+pbpaste > test/IRGen/Inputs/deserialize/clang-importer-witness-tables/regex.swift
```
Comment (slavapestov) [inline test/IRGen/Inputs/deserialize-clang-importer-witness-tables/clang-importer-witness-tables/regex.swift]: Did you mean to heck this in?

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@swift-ci Please test compiler performance

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@swift-ci Please test compiler performance

**Pull request merged**